core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceJoinOverConstantWithProject.java

```diff
@@ -0,0 +1,180 @@
+
+        switch (node.getType()) {
+            case INNER:
+                if (lefSourceSingleConstantRow) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+                break;
+            case LEFT:
+                if (lefSourceSingleConstantRow && isAtLeastScalar(right, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+                break;
+            case RIGHT:
+                if (lefSourceSingleConstantRow) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow && isAtLeastScalar(left, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+                break;
+            case FULL:
+                if (lefSourceSingleConstantRow && isAtLeastScalar(right, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(right, node.getRightOutputSymbols(), left, node.getLeftOutputSymbols(), context.getIdAllocator()));
+                }
+                if (rightSourceSingleConstantRow && isAtLeastScalar(left, context.getLookup())) {
+                    return Result.ofPlanNode(appendProjection(left, node.getLeftOutputSymbols(), right, node.getRightOutputSymbols(), context.getIdAllocator()));
+                }
+        }
+
+        return Result.empty();
+    }
+
+    private static boolean isUnconditional(JoinNode joinNode)
+    {
+        return joinNode.getCriteria().isEmpty() &&
+                (joinNode.getFilter().isEmpty() || joinNode.getFilter().get().equals(TRUE_LITERAL));
+    }
+
+    private boolean isSingleConstantRow(PlanNode node)
+    {
+        if (!(node instanceof ValuesNode)) {
+            return false;
+        }
+
+        ValuesNode values = (ValuesNode) node;
+        if (values.getRowCount() != 1) {
+            return false;
+        }
+
+        if (values.getOutputSymbols().isEmpty()) {
+            // handled by ReplaceRedundantJoinWithSource rule
+            return false;
+        }
```

Comment:
While convenient, this check doesn't belong in `isSingleConstantRow`. I hoist it outside of this method for clarity